sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSender.java

```diff
@@ -317,12 +317,13 @@ public String getHostname() {
     @Override
     public Mono<Integer> getLinkSize() {
         if (linkSize > 0) {
-            return Mono.just(this.linkSize);
+            // Using Mono.defer when returning cached linkSize, refer https://github.com/Azure/azure-sdk-for-java/issues/26372.
+            return Mono.defer(() -> Mono.just(this.linkSize));
```

Comment:
When the `Mono.defer` is subscribed to does it use the `linkSize` captured during the construction of the lambda or the value at the time of subscription?

Reply:
The lambda object captures `this` instance not the properties of `this`. The properties are read when lambda executes.

For example -

```java
private volatile int i = 0; 
public void foo() {     
    Mono<Integer> m = Mono.defer(() -> Mono.just(this.i));     
    this.i = 10;    
    System.out.println(m.block());  // outputs 10    
    this.i = 20;     
    System.out.println(m.block());  // outputs 20
 }
```

Side note: Also, in the ReactorSender, once linkSize is read from the link only once, it won't change once set.

Reply:
Capturing the offline discussion for future reference - for example, when we evaluate the potential places to replace _just_ with _defer_.

Why the code change is safe in this specific case -

```java
if (this.linkSize > 0) {   // Imperative check for cache value
    return Mono.defer(() -> Mono.just(this.linkSize));  // Reactive access to cached value
}
```

1. The `linkSize` can be set only once, so its value will not change between the Imperative check and Reactive (delayed) access.
2. The lambda in `Mono.defer` will not capture the `linkSize` property; it gets eagerly read (as a result of subscription).
3. The `linkSize` is volatile, hence accessing the cached value from a different thread (e.g., Project Reactor thread) will see the latest value (and won't get partial value).